crates/autopilot/src/run_loop.rs

```diff
@@ -318,59 +309,102 @@ impl RunLoop {
+        .fold(Vec::new(), |mut solutions, ((start, result), driver)| {
+            for solution in match result {
+                Ok(solutions) => {
+                    Metrics::get().solve_ok(driver, start);
+                    solutions
+                }
                 Err(err) => {
-                    tracing::warn!(?err, "driver solve error");
-                    None
+                    Metrics::get().solve_err(driver, start, &err);
+                    if matches!(err, SolveError::NoSolutions) {
+                        tracing::debug!(driver = %driver.url, "solver found no solution");
+                    } else {
+                        tracing::warn!(?err, driver = %driver.url, "solve error");
+                    }
+                    return solutions;
                 }
+            } {
+                match solution {
+                    Ok(solution) => {
+                        Metrics::get().solution_ok(driver);
+                        solutions.push(Participant { driver, solution })
+                    }
+                    Err(err) => {
+                        Metrics::get().solution_err(driver, &err);
+                        tracing::debug!(?err, driver = %driver.url, "invalid proposed solution");
+                    }
+                }
+            }
+            solutions
+        })
+    }
+
+    /// Computes a driver's solutions for the solver competition.
+    async fn solve(
+        &self,
+        driver: &Driver,
+        request: &solve::Request,
+    ) -> Result<Vec<Result<Solution, ZeroScoreError>>, SolveError> {
+        let response = tokio::time::timeout(SOLVE_TIME_LIMIT, driver.solve(request))
+            .await
+            .map_err(|_| SolveError::Timeout)?
+            .map_err(SolveError::Failure)?;
+        if response.solutions.is_empty() {
+            return Err(SolveError::NoSolutions);
+        }
+
+        Ok(response
+            .solutions
+            .into_iter()
+            .map(|solution| {
+                Ok(Solution {
+                    id: solution.solution_id,
+                    account: solution.submission_address,
+                    score: NonZeroU256::new(solution.score).ok_or(ZeroScoreError)?,
+                })
             })
-            .collect()
+            .collect())
     }
 
     /// Ask the winning solver to reveal their solution.
     async fn reveal(
         &self,
-        id: AuctionId,
-        solution_id: u64,
         driver: &Driver,
-    ) -> Result<reveal::Response> {
+        auction: AuctionId,
+        solution_id: u64,
+    ) -> Result<reveal::Response, RevealError> {
         let response = driver
             .reveal(&Request { solution_id })
             .await
-            .context("reveal")?;
-        ensure!(
-            response.calldata.internalized.ends_with(&id.to_be_bytes()),
-            "reveal auction id missmatch"
-        );
+            .map_err(RevealError::Failure)?;
+        if !response
```

Comment:
Interesting. Are we putting the auction ID data also on-chain?

Reply:
Yes, we started doing this.